runtime/vm/threadhelp.cpp

```diff
@@ -119,32 +118,73 @@ monitorWaitImpl(J9VMThread *vmThread, j9object_t object, I_64 millis, I_32 nanos
 	J9VMTHREAD_SET_BLOCKINGENTEROBJECT(vmThread, vmThread, object);
 	object = NULL;
 	internalReleaseVMAccessSetStatus(vmThread, thrstate);
+
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+	I_64 beforeWait = 0;
+	if (waitTimed) {
+		PORT_ACCESS_FROM_JAVAVM(javaVM);
+		beforeWait = j9time_nano_time();
+	}
+continueWait:
+#endif /* defined(J9VM_OPT_CRIU_SUPPORT) */
+
 	if (interruptable) {
 		rc = omrthread_monitor_wait_interruptable(monitor, millis, nanos);
 	} else {
 		rc = omrthread_monitor_wait_timed(monitor, millis, nanos);
 	}
+
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+	if (((J9THREAD_INTERRUPTED == rc)
+		|| (J9THREAD_PRIORITY_INTERRUPTED == rc))
+		&& J9_IS_SINGLE_THREAD_MODE(javaVM)
+	) {
+		if (waitTimed) {
+			PORT_ACCESS_FROM_JAVAVM(javaVM);
+			const I_32 oneMillion = 1000000;
+			I_64 waitedTimeNanos = j9time_nano_time() - beforeWait;
+			I_64 waitedTimeMillis = waitedTimeNanos / oneMillion;
+			if (waitedTimeMillis < millis) {
+				millis -= waitedTimeMillis;
+				waitedTimeNanos = (waitedTimeNanos % oneMillion);
+				if (waitedTimeNanos >= nanos) {
```

Comment:
what if user only specifies `wait(0, something)` ? doesnt the nanoSecond check need to be done independantly of the mili second?

Reply:
If less than `1ms` left for monitor waiting (either `wait(0, something)` or just the `millis` specified passed), I think there is a chance that this thread will be timed out and active while single thread mode is still on, hence waiting another `10ms` to accommodate such cases, the number was picked arbitrarily, I am open to a different timeout or nanos specified instead.